snakemake/script.py

```diff
@@ -309,7 +313,7 @@ def script(
     f = None
     try:
         path, source, language = get_source(path, basedir)
-        if language == "python":
+        if language == "python" or language == "jupyter":
```

Comment:
This is suboptimal, because jupyter is in principle language agnostic. Suggestion:
instead of keeping the script function as it is, convert it into a class, thereby splitting up the central parts into methods (get_preamble1, get_preamble2, write_script, execute_script, etc.). Then, inherit from that class in a new module `snakemake/jupyter.py`, where you redefine the write_script method into writing the notebook, and redefine the get_preamble2 method to add the jupyter specific code where you set the working directory. Finally, you redefine the execute_script method to perfom the jupyter specific execution.

I know, this is a lot to ask for. Hence, if you feel uncomfortable with that, I would also be fine to do the refactoring myself first, and then you can simply adapt to those changes.

Reply:
Ah yes, very good point! Since I mostly use the Python kernel I forgot about other options.
The refactoring sounds like a great idea. If you don't mind I can commit a rough draft, and only continue if it fits your vision.

To clarify: there would be a `ScriptExecutor` base class with methods:
* `get_preamble` (handling https://github.com/snakemake/snakemake/blob/0c1e56c163083034a6502e30b07dd7e5ef5ff67e/snakemake/script.py#L312-L479)
  * how would your suggested `get_preamble1` and `get_preamble2` differ?
* `write_script`(handling https://github.com/snakemake/snakemake/blob/0c1e56c163083034a6502e30b07dd7e5ef5ff67e/snakemake/script.py#L484-L504)
* `execute_script` (handling https://github.com/snakemake/snakemake/blob/0c1e56c163083034a6502e30b07dd7e5ef5ff67e/snakemake/script.py#L506-L562)

Would the end goal be for every script type (Python, R, Julia, ...) to be added to this framework, or only Jupyter?

Reply:
Every script type should be added to this framework.

Reply:
Perfect, I'll start with an implementation for Jupyter notebooks only and try to keep it general.